exec/vm.go

```diff
@@ -0,0 +1,320 @@
+			returns:        len(fn.Sig.ReturnTypes) != 0,
+		}
+	}
+
+	if module.Start != nil {
+		_, err := vm.ExecCode(int64(module.Start.Index))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return vm, nil
+}
+
+func (vm *VM) pushBool(v bool) {
+	if v {
+		vm.pushUint64(1)
+	} else {
+		vm.pushUint64(0)
+	}
+}
+
+func (vm *VM) fetchBool() bool {
+	return vm.fetchInt8() != 0
+}
+
+func (vm *VM) fetchInt8() int8 {
+	i := int8(vm.code[vm.pc])
+	vm.pc++
+	return i
+}
+
+func (vm *VM) fetchUint32() uint32 {
+	v := endianess.Uint32(vm.code[vm.pc:])
+	vm.pc += 4
+	return v
+}
+
+func (vm *VM) fetchInt32() int32 {
+	return int32(vm.fetchUint32())
+}
+
+func (vm *VM) fetchFloat32() float32 {
+	return math.Float32frombits(vm.fetchUint32())
+}
+
+func (vm *VM) fetchUint64() uint64 {
+	v := endianess.Uint64(vm.code[vm.pc:])
+	vm.pc += 8
+	return v
+}
+
+func (vm *VM) fetchInt64() int64 {
+	return int64(vm.fetchUint64())
+}
+
+func (vm *VM) fetchFloat64() float64 {
+	return math.Float64frombits(vm.fetchUint64())
+}
+
+func (vm *VM) popUint64() uint64 {
+	i := vm.stack[vm.stackTop]
+	vm.stackTop--
+	return i
+}
+
+func (vm *VM) popInt64() int64 {
+	return int64(vm.popUint64())
+}
+
+func (vm *VM) popFloat64() float64 {
+	return math.Float64frombits(vm.popUint64())
+}
+
+func (vm *VM) popUint32() uint32 {
+	return uint32(vm.popUint64())
+}
+
+func (vm *VM) popInt32() int32 {
+	return int32(vm.popUint32())
+}
+
+func (vm *VM) popFloat32() float32 {
+	return math.Float32frombits(vm.popUint32())
+}
+
+func (vm *VM) pushUint64(i uint64) {
+	vm.stackTop++
+	vm.stack[vm.stackTop] = i
+}
+
+func (vm *VM) pushInt64(i int64) {
+	vm.pushUint64(uint64(i))
+}
+
+func (vm *VM) pushFloat64(f float64) {
+	vm.pushUint64(math.Float64bits(f))
+}
+
+func (vm *VM) pushUint32(i uint32) {
+	vm.pushUint64(uint64(i))
+}
+
+func (vm *VM) pushInt32(i int32) {
+	vm.pushUint64(uint64(i))
+}
+
+func (vm *VM) pushFloat32(f float32) {
+	vm.pushUint32(math.Float32bits(f))
+}
+
+func (vm *VM) ExecCode(fnIndex int64, args ...uint64) (interface{}, error) {
+	if int(fnIndex) > len(vm.compiledFuncs) {
+		return nil, InvalidFunctionIndexError(fnIndex)
+	}
+	if len(vm.module.GetFunction(int(fnIndex)).Sig.ParamTypes) != len(args) {
+		return nil, ErrInvalidArgumentCount
+	}
+	compiled := vm.compiledFuncs[fnIndex]
+	vm.stack = make([]uint64, compiled.maxDepth)
+	vm.locals = make([]uint64, compiled.totalLocalVars)
+	vm.pc = 0
+	vm.stackTop = -1
+	vm.code = compiled.code
+
+	for i, arg := range args {
+		vm.locals[i] = arg
+	}
+
+	var rtrn interface{}
+	res := vm.execCode(compiled)
+	if compiled.returns {
+		rtrnType := vm.module.GetFunction(int(fnIndex)).Sig.ReturnTypes[0]
+		switch rtrnType {
+		case wasm.ValueTypeI32:
+			rtrn = uint32(res)
+		case wasm.ValueTypeI64:
+			rtrn = uint64(res)
+		case wasm.ValueTypeF32:
+			rtrn = math.Float32frombits(uint32(res))
+		case wasm.ValueTypeF64:
+			rtrn = math.Float64frombits(uint64(res))
+		default:
+			return nil, InvalidReturnTypeError(int8(rtrnType))
```

Comment:
The underlying type of `InvalidReturnTypeError` is already `int8`, so converting `rtrnType` to `int8` first is unnecessary. You can simplify this to:

```Go
return nil, InvalidReturnTypeError(rtrnType)
```